### PR TITLE
feat(datahub-client): add java file emitter

### PR DESCRIPTION
**Pull request by MugdhaHardikar-GSLab**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+   * @param config
+   */
+  public FileEmitter(FileEmitterConfig config) {
+
+    this.config = config;
+    this.eventFormatter = this.config.getEventFormatter();
+
+    DefaultPrettyPrinter pp = new DefaultPrettyPrinter()
+        .withObjectIndenter(new DefaultIndenter("    ", DefaultIndenter.SYS_LF))
+        .withArrayIndenter(new DefaultIndenter("    ", DefaultIndenter.SYS_LF));
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: Would it be better to define private static constant for the string with spaces/tabs?

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+@Slf4j
+public class FileEmitter implements Emitter {
+
+  private final EventFormatter eventFormatter;
+  private final FileEmitterConfig config;
+  private final ObjectMapper objectMapper = new ObjectMapper()
+      .setSerializationInclusion(JsonInclude.Include.NON_NULL);
+  private final JacksonDataTemplateCodec dataTemplateCodec = new JacksonDataTemplateCodec(objectMapper.getFactory());
+
+  private BufferedWriter writter;
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: fix spelling of writer

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  public Future<MetadataWriteResponse> emit(@SuppressWarnings("rawtypes") MetadataChangeProposalWrapper mcpw,
+      Callback callback) throws IOException {
+    return emit(this.eventFormatter.convert(mcpw), callback);
+  }
+
+  @Override
+  public Future<MetadataWriteResponse> emit(MetadataChangeProposal mcp, Callback callback) throws IOException {
+    if (this.writter == null) {
+      log.error("Emitter is closed. MCP could not be written on closed emitter.");
+      String errorMsg = "Emit function called after close.";
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: Can we keep single error message in exception and log?

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      Future<MetadataWriteResponse> response = createFailFuture(errorMsg);
+      if (callback != null) {
+        callback.onFailure(new Exception(errorMsg));
+      }
+      return response;
+    }
+
+    String serializedMCP = dataTemplateCodec.mapToString(mcp.data());
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: prefix object variables with "this."

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      } catch (InterruptedException | ExecutionException e) {
+        log.warn("Callback could not be executed.", e);
+      }
+    }
+    return response;
+  }
+
+  @Override
+  public boolean testConnection() throws IOException, ExecutionException, InterruptedException {
+    throw new UnsupportedOperationException("testConnection not relavent for File Emitter");
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: fix spelling of relevant

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  @Override
+  public Future<MetadataWriteResponse> emit(List<UpsertAspectRequest> request, Callback callback) throws IOException {
+    throw new UnsupportedOperationException("UpsertAspectRequest cannot be sent to File");
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: this error message is somewhat strange. Why not say "not relevant for File Emitter" as in testConnection() ?

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+      @Override
+      public MetadataWriteResponse get() throws InterruptedException, ExecutionException {
+        return MetadataWriteResponse.builder().success(true).responseContent("MCP witten to File").build();
+      }
+
+      @Override
+      public MetadataWriteResponse get(long timeout, TimeUnit unit)
+          throws InterruptedException, ExecutionException, TimeoutException {
+        return this.get();
```
Comment (ksrinath) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java:138]: Don't we have to check the timeouts here?

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+      @Override
+      public boolean isDone() {
+        return true;
+      }
+    };
+
+  }
+
+  private Future<MetadataWriteResponse> createFailFuture(String message) {
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: should we call it createFailureFuture to keep consistent with createSuccessFuture?

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    return new Future<MetadataWriteResponse>() {
+
+      @Override
+      public boolean cancel(boolean mayInterruptIfRunning) {
+        return false;
+      }
+
+      @Override
+      public MetadataWriteResponse get() throws InterruptedException, ExecutionException {
+        return MetadataWriteResponse.builder().success(false).responseContent(message).build();
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java:165]: Should we return such response or throw ExecutionException instead?
(Maybe we can just keep consistent with other emitters or the interface contract...)

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitterConfig.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import datahub.event.EventFormatter;
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class FileEmitterConfig {
+  @Builder.Default
+  private final String fileName = null;
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitterConfig.java:12]: Shouldn't fileName be mandatory?

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/kafka/KafkaEmitter.java [hunk 11]
```diff
@@ -37,7 +37,7 @@ public class KafkaEmitter implements Emitter {
   private static final int ADMIN_CLIENT_TIMEOUT_MS = 5000;
 
   /**
-   * The default constructor, prefer using the `create` factory method.
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/kafka/KafkaEmitter.java:40]: Confirm if this is necessary change...

File: metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+  private static final String RESOURCE_DIR = "src/test/resources";
+  private static final String GOLDEN_FILE_DIR = RESOURCE_DIR + "/golden_files";
+
+  @Rule
+  public TemporaryFolder tempFolder = new TemporaryFolder();
+
+  @Test
+  public void testFileEmitter() throws URISyntaxException, IOException {
+
+    String goldenJson = GOLDEN_FILE_DIR + "/mysql_mces_golden.json";
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java]: Is it relevant that these are "mysql" mces? 
If not, can we remove "mysql" from the name?

File: metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    if (map1.length != map2.length) {
+      fail("Number of MCPs are not equal. " + map1.length + ":" + map2.length);
+    }
+    for (int i = 0; i < map1.length; i++) {
+      Assert.assertEquals(map1[i], map2[i]);
+    }
+
+  }
+
+  private ArrayList<MetadataChangeProposal> getMCPs(String fileName) throws StreamReadException, DatabindException,
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java]: Would it be better to keep returnType as List<MetadataChangeProposal> ? Why do we specifically need ArrayList?

File: metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    emitter.close();
+
+    this.assertEqualJsonFile(goldenJson, outputFile);
+
+  }
+
+  private void assertEqualJsonFile(String file1, String file2) throws StreamReadException, DatabindException,
+      IOException {
+    TypeReference<Map<String, Object>[]> typeRef = new TypeReference<Map<String, Object>[]>() {
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java]: Can we read the files as a List<Map<String,Object>> and do a .equals() on the list? If it works, it'll be simpler...

File: metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    for (Map<String, Object> map : maps) {
+      String json = objectMapper.writeValueAsString(map);
+      DataMap data = dataTemplateCodec.stringToMap(json);
+      mcps.add(new MetadataChangeProposal(data));
+    }
+    return mcps;
+  }
+
+  @Test
+  public void testSucessCallback() throws Exception {
```
Comment (ghost) [inline metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java]: fix spelling of Success

Comment (MugdhaHardikar-GSLab) [reply metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java]: Done

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java (hunk 1 above)
Comment (MugdhaHardikar-GSLab) [reply metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java]: Done

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/kafka/KafkaEmitter.java (hunk 11 above)
Comment (MugdhaHardikar-GSLab) [reply metadata-integration/java/datahub-client/src/main/java/datahub/client/kafka/KafkaEmitter.java:40]: Yes. As there is no create method in kafka emitter

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java (hunk 9 above)
Comment (MugdhaHardikar-GSLab) [reply metadata-integration/java/datahub-client/src/main/java/datahub/client/file/FileEmitter.java:165]: This is mainly to be in sync with other emitters.

File: metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java (hunk 14 above)
Comment (MugdhaHardikar-GSLab) [reply metadata-integration/java/datahub-client/src/test/java/datahub/client/file/FileEmitterTest.java]: Done

**Pull request closed without merging**